File: clients/dart/amenolink/lib/src/interfaces.dart
import 'dart:async';
import 'dtos.dart';

typedef ActionHandler<T, R> = FutureOr<R> Function(T input);
typedef CacheAllHandler = void Function(String key, dynamic value);
typedef CacheKeyHandler<T> = void Function(T? value);
typedef TopicHandler<T> = void Function(TopicMessage<T> message);

abstract interface class Action {
  String get name;
  Future<T> request<T>(dynamic payload);
  Future<void> queue(dynamic payload);
}

abstract interface class CacheWatcher {
  String get group;
  void all(CacheAllHandler handler);
  void key<T>(String key, CacheKeyHandler<T> handler);
  void dispose();
}

abstract interface class Cache {
  String get group;
  CacheWatcher watch();
  Future<T?> get<T>(String key);
  Future<void> set(String key, dynamic value);
  Future<T> getOrCreate<T>(String key, Future<T> Function() creator);
  Future<Map<String, dynamic>> all();
  Future<void> clear();
  Future<void> delete(String key);
}

abstract interface class Topic<T> {
  String get name;
  void subscribe(TopicHandler<T> handler);
  Future<void> publish(T? value, {Message? previous});
  void dispose();
}

abstract interface class ActionContext {
  ActionRequest get request;
  void log(String message);
}

abstract interface class ActionRouter {
  void add<T, R>({required String route, required ActionHandler<T, R> handler, T Function(dynamic json)? parseInput});
  void serve();
}

File: clients/csharp/AmenoLink/Interfaces.cs
namespace AmenoLink;

public interface IAction
{
    string Name { get; }
    Task<T> Request<T>(object? payload = null);
    Task Queue(object? payload = null);
}

public interface ICacheWatcher : IDisposable
{
    string Group { get; }
    void All(Action<string, object?> handler);
    void Key<T>(string key, Action<T?> handler);
}

public interface ICache
{
    string Group { get; }
    ICacheWatcher Watch();
    Task<T?> Get<T>(string key);
    Task Set<T>(string key, T value);
    Task<T> GetOrCreate<T>(string key, Func<Task<T>> creator);
    Task<T> GetOrCreate<T>(string key, Func<T> creator);
    Task<Dictionary<string, object?>> All();
    Task Clear();
    Task Delete(string key);
}

public interface ITopic<T> : IDisposable
{
    string Name { get; }
    void Subscribe(Func<TopicMessage<T>, Task> handler);
    Task Publish(T value, Message? previous = null);
}

public interface IActionContext
{
    ActionRequest<object?> Request { get; }
    void Log(string message);
}

public interface IActionRouter
{
    void Add<T, R>(string route, Func<T, Task<R>> handler);
    void Add<T, R>(string route, Func<T, R> handler);
    void Serve();
}

public interface IConnection
{
    void Subscribe(Action<ConnectionStatus> listener);
    void Unsubscribe(Action<ConnectionStatus> listener);
    void UnsubscribeAll();
    Task Connect();
    Task Disconnect();
}

internal interface IConnectionManager
{
    bool IsConnected { get; }
    Task Send(string method, object? arg1 = null, CancellationToken cancellationToken = default);
    void SubscribeTopic(ITopicSubscriber topicInstance);
    void UnsubscribeTopic(ITopicSubscriber topicInstance);
    void SubscribeWatcher(ICacheWatcherSubscriber watcherInstance);
    void UnsubscribeWatcher(ICacheWatcherSubscriber watcherInstance);
}

internal interface ITopicSubscriber
{
    string Name { get; }
    void DispatchMessage(TopicMessage<object> message);
}

internal interface ICacheWatcherSubscriber
{
    string Group { get; }
    void DispatchValueChanged(string key, object? value);
}
